lib/client.js

```diff
@@ -1474,6 +1515,155 @@ function write (client, request) {
+  // expect it can cause undefined behavior on some
+  // servers and corrupt connection state. Do not
+  // re-use the connection for further requests.
+
+
+  const expectsPayload = (
+    method === 'PUT' ||
+    method === 'POST' ||
+    method === 'PATCH'
+  )
+
+  if (body && typeof body.read === 'function') {
+    // Try to read EOF in order to get length.
+    body.read(0)
+  }
+
+  let contentLength = util.bodyLength(body)
+
+  if (contentLength == null) {
+    contentLength = request.contentLength
+  }
+
+  if (contentLength === 0 || !expectsPayload) {
+    // https://tools.ietf.org/html/rfc7230#section-3.3.2
+    // A user agent SHOULD NOT send a Content-Length header field when
+    // the request message does not contain a payload body and the method
+    // semantics do not anticipate such a body.
+
+    contentLength = null
+  }
+
+  if (request.contentLength != null && request.contentLength !== contentLength) {
+    if (client[kStrictContentLength]) {
+      errorRequest(client, request, new RequestContentLengthMismatchError())
+      return false
+    }
+
+    process.emitWarning(new RequestContentLengthMismatchError())
+  }
+
+  try {
+    // TODO(HTTP/2): Should we call onConnect immediately or on stream ready event?
+    request.onConnect((err) => {
+      if (request.aborted || request.completed) {
+        return
+      }
+
+      errorRequest(client, request, err || new RequestAbortedError())
+    })
+  } catch (err) {
+    errorRequest(client, request, err)
+  }
+
+  if (request.aborted) {
+    return false
+  }
+
+  const headers = { ...request.headers }
+  headers[':authority'] = host || client[kHost]
+  headers[':path'] = path
+
+  // TODO(HTTP/2): Expect: 100-continue
+
+  /* istanbul ignore else: assertion */
+  if (!body) {
+    if (contentLength === 0) {
+      headers['content-length'] = '0'
+    } else {
+      assert(contentLength == null, 'no body must not have content length')
+    }
+  } else if (util.isBuffer(body)) {
+    assert(contentLength === body.byteLength, 'buffer body must have content length')
+
+    headers['content-length'] = String(contentLength)
+
+    process.nextTick(() => {
+      stream.end(body)
+      request.onBodySent(body)
+    })
+  } else if (util.isBlob(body)) {
+    process.nextTick(() => {
+      writeBlob({ client, request, stream, contentLength, expectsPayload })
+    })
+  } else if (util.isStream(body)) {
+    process.nextTick(() => {
+      writeStream({ client, request, stream, contentLength, expectsPayload })
+    })
+  } else if (util.isIterable(body)) {
+    process.nextTick(() => {
+      writeIterable({ client, request, stream, contentLength, expectsPayload })
+    })
+  } else {
+    assert(false)
+  }
+
+  console.log('http/2 request')
+  // TODO(HTTP/2): ref only if current streams count is 0
+  session.ref()
+  // TODO(HTTP/2): The handler expects an array but the native http2 module provides us with an object. What should we do?
```

Comment:
It is safe to assume here that we should just transform the array of headers into objects before triggering the request?

Reply:
alternatively, allow the handler to accept an object, rather than having to transfer to <-> from multiple times?

Reply:
I think already does, within the handler, unless I overlooked?